compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -749,6 +756,79 @@ public void visit(BLangErrorType errorType) {
             dlog.error(errorType.detailType.pos, DiagnosticErrorCode.INVALID_ERROR_DETAIL_TYPE, errorType.detailType,
                     symTable.detailType);
         }
+        analyzeDef(errorType.detailType, env);
+    }
+
+    @Override
+    public void visit(BLangConstrainedType constrainedType) {
+        analyzeDef(constrainedType.constraint, env);
+    }
+
+    @Override
+    public void visit(BLangFunctionTypeNode functionTypeNode) {
+        List<BLangVariable> params = functionTypeNode.params;
+        for (BLangVariable param : params) {
+            analyzeDef(param.typeNode, env);
+        }
+        if (functionTypeNode.restParam != null) {
+            analyzeDef(functionTypeNode.restParam.typeNode, env);
+        }
+        if (functionTypeNode.returnTypeNode != null) {
+            analyzeDef(functionTypeNode.returnTypeNode, env);
+        }
+    }
+
+    @Override
+    public void visit(BLangUnionTypeNode unionTypeNode) {
+        for (BLangType memberType : unionTypeNode.memberTypeNodes) {
+            analyzeDef(memberType, env);
+        }
+    }
+
+    @Override
+    public void visit(BLangStreamType streamType) {
+        analyzeNode(streamType.constraint, env);
+        if (streamType.error != null) {
+            analyzeDef(streamType.error, env);
```

Comment:
Handle the null check inside analyze function so that we don't need to repeat the check

Reply:
analyze function returns a `BType`. if we were to handle the null check there we'll have to return something for it too.